File: apps/ios/Loonext/Features/Settings/EmergencyCard.swift
import SwiftUI

/// #460 — the emergency words and reply, in the owner's hands.
///
/// The shipped defaults told every customer of every workspace to reply URGENT
/// "for a no-heat or burst-pipe emergency", and answered them with a sentence
/// about smelling gas and calling the utility. Landscapers, locksmiths and
/// mobile mechanics all sent that. Now the words and the reply are theirs.
///
/// WHY THIS IS ITS OWN CARD rather than three more rows on the away card: that
/// one already carries three decisions (reply on/off, the message, and whether
/// an emergency word wakes the crew). Adding the word list and the reply makes
/// five, past what a reader holds at once, and the two halves answer different
/// questions — "what do we say when we're shut" versus "what counts as an
/// emergency and what goes back".
///
/// It sits DIRECTLY beneath the away card and nowhere else, because the away
/// message is the sentence that tells a customer the word. Same copy as web and
/// Android, deliberately: a rule worded three ways is three rules.
@MainActor
struct EmergencyCard: View {
    let scope: SettingsScope
    let company: CompanyView
    let onCompanyUpdated: @MainActor (CompanyView) -> Void

    @State private var words: [String]
    @State private var draft = ""
    @State private var message: String
    @State private var replyEnabled: Bool
    @State private var saving = false
    @State private var error: String?

    @Environment(\.appLocale) private var appLocale

    init(
        scope: SettingsScope,
        company: CompanyView,
        onCompanyUpdated: @escaping @MainActor (CompanyView) -> Void
    ) {
        self.scope = scope
        self.company = company
        self.onCompanyUpdated = onCompanyUpdated
        // Seeded from the EFFECTIVE list, never the raw column. An owner who has
        // never opened this screen has a nil column, and an empty box would read
        // as "nothing is watched for" — the opposite of the truth, and the
        // fastest way to make somebody think the feature is broken.
        _words = State(initialValue: company.effectiveEmergencyWords)
        _message = State(initialValue: company.emergency_message ?? "")
        // #553: held locally so the switch moves on the tap rather than a round
        // trip later.
        _replyEnabled = State(initialValue: company.emergency_reply_enabled)
    }

    private func t(_ key: String, _ vars: [String: String] = [:]) -> String {
        AppStrings.translate(appLocale, key, vars)
    }

    /// The sentence no setting removes, in the reader's language — resolved
    /// here rather than held as a constant, exactly as `EmergencyCard.kt` does.
    private var safetyLine: String { t(emergencySafetyLineKey) }

    private var canEdit: Bool { SettingsRoleGate.canEditWorkspace(scope.role) }
    private var savedWords: [String] { company.effectiveEmergencyWords }
    private var trimmedMessage: String {
        message.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private var dirty: Bool {
        words != savedWords
            || trimmedMessage
                != (company.emergency_message ?? "")
                    .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// The composed preview: the owner's body (or the product default) followed
    /// by the sentence no setting removes. The server's composed message is the
    /// truth whenever nothing is unsaved, which is why it shows when empty.
    private var preview: String {
        let body = trimmedMessage.isEmpty
            ? company.emergency_effective_message
            : trimmedMessage
        return body.contains(safetyLine)
            ? body
            : "\(body) \(safetyLine)"
    }

    var body: some View {
        SettingsCard(
            title: t("settings.emergencyTitle"),
            description: t("settings.emergencyIntro")
        ) {
            Text(t("settings.emergencyWordsHeading"))
                .font(.callout)
            Text(t("settings.emergencyWordsHelp"))
                .font(.footnote)
                .foregroundStyle(.secondary)

            Spacer().frame(height: 8)
            // The word IS the content, so it stays legible and the remove
            // affordance rides beside it rather than replacing it.
            WrappingWords(words: words, canEdit: canEdit && !saving, onRemove: remove)

            if canEdit {
                Spacer().frame(height: 8)
                HStack(spacing: 8) {
                    TextField(t("settings.emergencyAddWordPlaceholder"), text: $draft)
                        .textFieldStyle(.roundedBorder)
                        .autocorrectionDisabled()
                        .textInputAutocapitalization(.characters)
                        .disabled(saving)
                        .onChange(of: draft) { _, value in
                            if value.count > 15 { draft = String(value.prefix(15)) }
                        }
                    Button(t("settings.emergencyAddWordAction")) { add() }
                        .disabled(saving)
                }
            }
            if !company.emergency_keywords_are_custom {
                Text(t("settings.emergencyDefaults"))
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }

            Spacer().frame(height: 16)
            Text(t("settings.emergencyReplyHeading"))
                .font(.callout)
            // #553: the switch lives NEXT TO the thing it governs.
            //
            // It always existed, but under Hours — a screen away from the words and
            // the message it controls — so the founder reported the reply as
            // "enforced". A setting nobody can find is not a setting.
            //
            // And it is now a DIFFERENT switch. `emergency_keyword_enabled` used to
            // gate four things at once, so the only way to stop us texting on the
            // crew's behalf was to stop the product noticing emergencies. Turning
            // this one off keeps the crew escalation, the push and the inbox flag,
            // and withholds only the message.
            Toggle(t("settings.emergencyTextBack"), isOn: $replyEnabled)
                .font(.callout)
                .disabled(!canEdit || saving)
                .onChange(of: replyEnabled) { _, next in saveReplyEnabled(next) }
                .padding(.top, 4)
            Text(t("settings.emergencyTextBackHelp"))
                .font(.footnote)
                .foregroundStyle(.secondary)
            Spacer().frame(height: 6)
            Text(t("settings.emergencyReplyHelp"))
                .font(.footnote)
                .foregroundStyle(.secondary)

            Spacer().frame(height: 6)
            TextField(
                company.emergency_effective_message,
                text: $message,
                axis: .vertical
            )
            .lineLimit(3...6)
            .textFieldStyle(.roundedBorder)
            .disabled(!canEdit || saving)
            .onChange(of: message) { _, value in
                if value.count > 1000 { message = String(value.prefix(1000)) }
            }
            Text(
                t(
                    company.emergency_message_is_custom
                        ? "settings.emergencyCount"
                        : "settings.emergencyCountDefault",
                    ["count": String(message.count)]
                )
            )
            .font(.caption2)
            .foregroundStyle(.secondary)

            // The honest part of this screen. An owner editing the body needs to
            // see that one sentence follows it whatever they write — otherwise
            // they will believe they removed it, and find out from a customer.
            Spacer().frame(height: 12)
            PreviewBubble(label: t("settings.emergencyPreviewLabel"), text: preview)
            Text(t("settings.emergencySafetyLineNote", ["line": safetyLine]))
                .font(.caption2)
                .foregroundStyle(.secondary)

            InlineError(error)

            if canEdit {
                if dirty {
                    Spacer().frame(height: 10)
                    Button(
                        saving ? t("common.saving") : t("settings.emergencySaveAction")
                    ) { save() }
                        .buttonStyle(.borderedProminent)
                        .tint(BrandColor.olive)
                        .disabled(saving)
                }
            } else {
                Spacer().frame(height: 4)
                ReadOnlyLine(t("settings.emergencyReadOnly"))
            }
        }
    }

    private func add() {
        let raw = draft.trimmingCharacters(in: .whitespaces)
        if let problem = emergencyKeywordError(raw, locale: appLocale) {
            error = problem
            return
        }
        let word = raw.uppercased()
        if words.contains(word) {
            error = t("settings.emergencyDuplicateWord", ["word": word])
            return
        }
        if words.count >= 10 {
            error = t("settings.emergencyTooManyWords")
            return
        }
        error = nil
        words.append(word)
        draft = ""
    }

    private func remove(_ word: String) {
        // Never down to zero. An empty list is not "no emergencies" — the switch
        // on the away card says that, honestly and reversibly. Watching for
        // nothing while the switch reads ON is the #414 defect.
        if words.count == 1 {
            error = t("settings.emergencyKeepOneWord")
            return
        }
        error = nil
        words.removeAll { $0 == word }
    }

    /// #553: saved on the tap, not behind the card's Save button.
    ///
    /// The words and the message are a draft somebody composes and commits. This is
    /// a switch, and a switch that needs a separate Save is a switch people believe
    /// they have already thrown. On failure it goes back to what the server has, so
    /// the control never shows a state the account is not in.
    private func saveReplyEnabled(_ next: Bool) {
        Task {
            do {
                let updated = try await scope.repo.updateCompany(
                    scope.companyId,
                    patch: .object(["emergency_reply_enabled": .bool(next)])
                )
                onCompanyUpdated(updated)
            } catch {
                replyEnabled = !next
                self.error = error.userMessage
            }
        }
    }

    private func save() {
        error = nil
        saving = true
        var fields: [String: JSONValue] = [:]
        // Only send the list when it is genuinely the owner's. Echoing the
        // product defaults back would FREEZE them on this workspace, so
        // improving them later would never reach it.
        if words != savedWords || company.emergency_keywords_are_custom {
            fields["emergency_keywords"] = .array(words.map { .string($0) })
        }
        fields["emergency_message"] =
            trimmedMessage.isEmpty ? .null : .string(trimmedMessage)
        let body = JSONValue.object(fields)
        Task {
            do {
                let updated = try await scope.repo.updateCompany(scope.companyId, patch: body)
                onCompanyUpdated(updated)
                scope.showMessage(t("settings.emergencySaved"))
            } catch {
                self.error = error.userMessage
            }
            saving = false
        }
    }
}

/// The one sentence appended to every emergency reply, mirroring
/// `EMERGENCY_SAFETY_LINE` in shared. Read from the catalogue because this
/// screen has to PREVIEW the composed message while the owner is still typing —
/// the server's composed value is a round trip behind.
///
/// BOTH LANGUAGES ARE THE SERVER'S OWN, copied character for character from
/// `packages/shared/src/locale.ts`. This is the one sentence in the product
/// with a SAFETY property: everything else degrades to "the reader gets
/// English" when a translation is missing, and this would degrade to somebody
/// in danger being told what to do in a language they may not read. A prettier
/// French written here would preview a sentence the server never sends, which
/// is the same failure with better spelling. 911 is the number in Canada and
/// the US alike, so it is as region-neutral in French as it is in English.
let emergencySafetyLineKey = "settings.emergencySafetyLine"

/// The keyword chips, wrapped.
///
/// `Layout`-free on purpose: a handful of short words never needs a custom
/// layout pass, and `LazyVGrid` with adaptive columns would give every chip the
/// same width — which for words of very different lengths reads as a table
/// rather than as a list of words.
@MainActor
private struct WrappingWords: View {
    let words: [String]
    let canEdit: Bool
    let onRemove: (String) -> Void

    @Environment(\.appLocale) private var appLocale

    private func t(_ key: String, _ vars: [String: String] = [:]) -> String {
        AppStrings.translate(appLocale, key, vars)
    }

    var body: some View {
        FlexibleRows(items: words) { word in
            Button {
                if canEdit { onRemove(word) }
            } label: {
                Text(canEdit ? t("settings.emergencyWordChip", ["word": word]) : word)
                    .font(.system(.footnote, design: .monospaced))
                    .padding(.horizontal, 10)
                    .padding(.vertical, 6)
                    // `insetDeep`, not `paper`: this chip sits ON a card, and
                    // a paper chip on a paper card is invisible in both themes.
                    // The inset ladder is what the palette provides for a well
                    // sitting inside a raised surface.
                    .background(BrandColor.insetDeep, in: Capsule())
                    .foregroundStyle(BrandColor.ink)
            }
            .buttonStyle(.plain)
            .disabled(!canEdit)
            .accessibilityLabel(
                canEdit ? t("settings.emergencyRemoveWordLabel", ["word": word]) : word
            )
        }
    }
}

/// Rows that wrap, built by chunking rather than measuring.
///
/// Four words per row at the shortest phone width, which every keyword fits
/// inside because the column refuses anything over 15 characters. Chunking is
/// exact where measurement would be approximate, and the bound is enforced by
/// the database rather than hoped for here.
@MainActor
private struct FlexibleRows<Item: Hashable, Content: View>: View {
    let items: [Item]
    @ViewBuilder let content: (Item) -> Content

    private var rows: [[Item]] {
        stride(from: 0, to: items.count, by: 3).map { start in
            Array(items[start..<min(start + 3, items.count)])
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(rows.enumerated()), id: \.offset) { row in
                HStack(spacing: 6) {
                    ForEach(row.element, id: \.self) { content($0) }
                    Spacer(minLength: 0)
                }
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/EmergencyCard.kt
package com.loonext.android.features.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.AssistChip
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.core.model.CompanyView
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * #460 — the emergency words and reply, in the owner's hands.
 *
 * WHY THIS IS ITS OWN CARD rather than three more rows on the away card: that
 * one already carries three decisions (reply on/off, the message, and whether an
 * emergency word wakes the crew). Adding the word list and the reply makes five,
 * past what a reader holds at once, and the two halves answer different
 * questions — "what do we say when we're shut" versus "what counts as an
 * emergency and what goes back". Splitting them is what lets each stay short.
 *
 * It sits DIRECTLY beneath the away card and nowhere else, because the away
 * message is the sentence that tells a customer the word. Same copy as web and
 * iOS, deliberately: a rule worded three ways is three rules.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun EmergencyCard(
    scope: SettingsScope,
    company: CompanyView,
    onCompanyUpdated: (CompanyView) -> Unit,
) {
    val canEdit = SettingsRoleGate.canEditWorkspace(scope.role)
    val coroutines = rememberCoroutineScope()

    val savedWords = company.effectiveEmergencyWords
    // Seeded from the EFFECTIVE list, never the raw column. An owner who has
    // never opened this screen has a null column, and an empty box would read as
    // "nothing is watched for" — the opposite of the truth, and the fastest way
    // to make somebody think the feature is broken.
    var words by remember(savedWords) { mutableStateOf(savedWords) }
    var draft by remember { mutableStateOf("") }
    var message by remember(company.emergency_message) {
        mutableStateOf(company.emergency_message.orEmpty())
    }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    // #553: held locally so the switch moves on the tap rather than a round trip
    // later, re-seeded whenever the company row changes.
    var replyEnabled by remember(company.emergency_reply_enabled) {
        mutableStateOf(company.emergency_reply_enabled)
    }

    /**
     * #553: saved on the tap, not behind the card's Save button.
     *
     * The words and the message are a draft somebody composes and commits. This is
     * a switch, and a switch that needs a separate Save is a switch people believe
     * they have already thrown. On failure it goes back to what the server has, so
     * the control never shows a state the account is not in.
     */
    fun saveReplyEnabled(next: Boolean) {
        coroutines.launch {
            try {
                val updated = scope.repo.updateCompany(
                    scope.companyId,
                    buildJsonObject { put("emergency_reply_enabled", next) },
                )
                onCompanyUpdated(updated)
            } catch (cause: Exception) {
                replyEnabled = !next
                error = cause.userMessage()
            }
        }
    }
    val trimmedMessage = message.trim()
    val dirty = words != savedWords ||
        trimmedMessage != company.emergency_message.orEmpty().trim()

    // The composed preview: the owner's body (or the product default) followed
    // by the sentence no setting removes. The server's composed message is the
    // truth whenever nothing is unsaved, which is why it shows when empty.
    val previewBody = trimmedMessage.ifEmpty { company.emergency_effective_message }
    val preview =
        if (previewBody.contains(EMERGENCY_SAFETY_LINE)) previewBody
        else "$previewBody $EMERGENCY_SAFETY_LINE"

    // `addWord` and the chip's onClick both run outside composition, and one of
    // them has to name the offending word — so the locale is captured here and
    // the sentence is built where it is needed rather than being read from a
    // @Composable that is no longer running.
    val locale = LocalAppLocale.current
    val tooManyWords = t("settings.emergencyTooManyWords")
    val keepOneWord = t("settings.emergencyKeepOneWord")
    val savedMessage = t("settings.emergencySaved")

    fun addWord() {
        val raw = draft.trim()
        val problem = emergencyKeywordError(raw)
        if (problem != null) {
            error = problem
            return
        }
        val word = raw.uppercase()
        when {
            word in words -> error = AppStrings.translate(
                locale,
                "settings.emergencyDuplicateWord",
                mapOf("word" to word),
            )
            words.size >= 10 -> error = tooManyWords
            else -> {
                error = null
                words = words + word
                draft = ""
            }
        }
    }

    SettingsCard(
        title = t("settings.emergencyTitle"),
        description = t("settings.emergencyIntro"),
    ) {
        Text(
            t("settings.emergencyWordsHeading"),
            style = MaterialTheme.typography.labelLarge,
        )
        Text(
            t("settings.emergencyWordsHelp"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 2.dp),
        )
        FlowRow(
            Modifier.fillMaxWidth().padding(top = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            words.forEach { word ->
                AssistChip(
                    onClick = {
                        // Never down to zero. An empty list is not "no
                        // emergencies" — the switch on the away card says that,
                        // honestly and reversibly. Watching for nothing while
                        // the switch reads ON is the #414 defect.
                        if (!canEdit || saving) return@AssistChip
                        if (words.size == 1) {
                            error = keepOneWord
                        } else {
                            error = null
                            words = words - word
                        }
                    },
                    // The word IS the content, so it stays legible and the
                    // remove affordance rides beside it rather than replacing
                    // it. Text rather than an icon: at chip size a glyph next
                    // to a monospaced word reads as part of the word.
                    label = {
                        Text(
                            if (canEdit) {
                                t("settings.emergencyWordChip", "word" to word)
                            } else {
                                word
                            },
                            fontFamily = FontFamily.Monospace,
                        )
                    },
                )
            }
        }
        if (canEdit) {
            Row(
                Modifier.fillMaxWidth().padding(top = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                OutlinedTextField(
                    value = draft,
                    onValueChange = { if (it.length <= 15) draft = it },
                    modifier = Modifier.weight(1f),
                    singleLine = true,
                    enabled = !saving,
                    label = { Text(t("settings.emergencyAddWordLabel")) },
                    placeholder = { Text(t("settings.emergencyAddWordPlaceholder")) },
                )
                Spacer(Modifier.width(8.dp))
                OutlinedButton(enabled = !saving, onClick = { addWord() }) {
                    Text(t("settings.emergencyAddWordAction"))
                }
            }
        }
        if (!company.emergency_keywords_are_custom) {
            Text(
                t("settings.emergencyDefaults"),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 4.dp),
            )
        }

        Spacer(Modifier.height(16.dp))
        Text(t("settings.emergencyReplyHeading"), style = MaterialTheme.typography.labelLarge)
        // #553: the switch lives NEXT TO the thing it governs.
        //
        // It always existed, but under Hours — a screen away from the words and the
        // message it controls — so the founder reported the reply as "enforced". A
        // setting nobody can find is not a setting.
        //
        // And it is now a DIFFERENT switch. `emergency_keyword_enabled` used to
        // gate four things at once, so the only way to stop us texting on the
        // crew's behalf was to stop the product noticing emergencies. Turning this
        // one off keeps the crew escalation, the push and the inbox flag, and
        // withholds only the message.
        LabeledSwitchRow(
            label = t("settings.emergencyTextBack"),
            checked = replyEnabled,
            onCheckedChange = { next ->
                replyEnabled = next
                saveReplyEnabled(next)
            },
            supporting = t("settings.emergencyTextBackHelp"),
            modifier = Modifier.padding(top = 4.dp),
        )
        Text(
            t("settings.emergencyReplyHelp"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 2.dp),
        )
        OutlinedTextField(
            value = message,
            onValueChange = { if (it.length <= 1000) message = it },
            modifier = Modifier.fillMaxWidth().padding(top = 6.dp),
            minLines = 3,
            enabled = canEdit && !saving,
            placeholder = { Text(company.emergency_effective_message) },
            supportingText = {
                Text(
                    if (company.emergency_message_is_custom) {
                        t("settings.emergencyCount", "count" to message.length.toString())
                    } else {
                        t(
                            "settings.emergencyCountDefault",
                            "count" to message.length.toString(),
                        )
                    },
                )
            },
        )

        // The honest part of this screen. An owner editing the body needs to see
        // that one sentence follows it whatever they write — otherwise they will
        // believe they removed it, and find out from a customer.
        Spacer(Modifier.height(12.dp))
        PreviewBubble(label = t("settings.emergencyPreviewLabel"), text = preview)
        Text(
            t("settings.emergencySafetyLineNote", "line" to EMERGENCY_SAFETY_LINE),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 4.dp),
        )

        InlineError(error)

        if (canEdit && dirty) {
            Button(
                enabled = !saving,
                modifier = Modifier.padding(top = 10.dp),
                onClick = {
                    error = null
                    saving = true
                    coroutines.launch {
                        try {
                            val body = buildJsonObject {
                                // Only send the list when it is genuinely the
                                // owner's. Echoing the product defaults back
                                // would FREEZE them on this workspace, so
                                // improving them later would never reach it.
                                if (words != savedWords || company.emergency_keywords_are_custom) {
                                    put(
                                        "emergency_keywords",
                                        buildJsonArray { words.forEach { add(it) } },
                                    )
                                }
                                if (trimmedMessage.isEmpty()) {
                                    put("emergency_message", JsonNull)
                                } else {
                                    put("emergency_message", trimmedMessage)
                                }
                            }
                            val updated = scope.repo.updateCompany(scope.companyId, body)
                            onCompanyUpdated(updated)
                            scope.showMessage(savedMessage)
                        } catch (cause: Exception) {
                            error = cause.userMessage()
                        } finally {
                            saving = false
                        }
                    }
                },
            ) {
                Text(
                    if (saving) t("common.saving") else t("settings.emergencySaveAction"),
                )
            }
        } else if (!canEdit) {
            Spacer(Modifier.height(4.dp))
            ReadOnlyLine(t("settings.emergencyReadOnly"))
        }
    }
}

/**
 * The one sentence appended to every emergency reply, mirroring
 * `EMERGENCY_SAFETY_LINE` in shared. Kept in Kotlin because this screen has to
 * PREVIEW the composed message while the owner is still typing — the server's
 * composed value is a round trip behind.
 */
const val EMERGENCY_SAFETY_LINE = "If anyone is in danger, call 911."
